### PR TITLE
optimize upstream mixer check and report call configuration

### PR DESCRIPTION
**Pull request by ayj**
* disable mixer check and report calls for outbound routes/services
* omit API, Quota, and Auth specs from service config when mixer check
   and report calls are disabled.

fixes https://github.com/istio/istio/issues/2418

### CODE REVIEW

**Comment by ldemailly** [comment]
we should possibly still do checks for out of mesh calls but for now this is a better state vs the 2x perf regression

File: pilot/proxy/envoy/mixer.go [hunk 1]
```diff
@@ -145,7 +145,7 @@ func buildMixerOpaqueConfig(check, forward bool, destinationService string) map[
 
 // Mixer filter uses outbound configuration by default (forward attributes,
 // but not invoke check calls)
-func mixerHTTPRouteConfig(role proxy.Node, instances []*model.ServiceInstance, config model.IstioConfigStore) *FilterMixerConfig { // nolint: lll
+func mixerHTTPRouteConfig(role proxy.Node, instances []*model.ServiceInstance, outboundRoute bool, config model.IstioConfigStore) *FilterMixerConfig { // nolint: lll
```
Comment (qiwzhang) [inline pilot/proxy/envoy/mixer.go]: Will this approach work for Ingress proxy and Egress proxy.

As I understand,  we need to call Mixer for Ingress and Egress proxy?   You can double check with @kyessenov and @mandarjog

File: pilot/proxy/envoy/mixer.go [hunk 2]
```diff
@@ -201,13 +201,13 @@ func mixerHTTPRouteConfig(role proxy.Node, instances []*model.ServiceInstance, o
 					},
 				},
 			},
-			DisableCheckCalls:  outboundRoute,
-			DisableReportCalls: outboundRoute,
+			DisableCheckCalls:  outboundRoute || mesh.DisablePolicyChecks,
+			DisableReportCalls: outboundRoute || role.Type == proxy.Ingress,
```
Comment (qiwzhang) [inline pilot/proxy/envoy/mixer.go]: We may need to send Report for Egress too.

Comment (ayj) [reply pilot/proxy/envoy/mixer.go]: I don't see an easy way to do this with the current code. It should be easier to do once we switch over to per-route config.

Comment (qiwzhang) [reply pilot/proxy/envoy/mixer.go]: OK. then.

File: pilot/proxy/envoy/mixer.go [hunk 3]
```diff
@@ -201,42 +201,48 @@ func mixerHTTPRouteConfig(role proxy.Node, instances []*model.ServiceInstance, c
 					},
 				},
 			},
+			DisableCheckCalls:  outboundRoute || mesh.DisablePolicyChecks,
+			DisableReportCalls: outboundRoute || role.Type == proxy.Ingress,
```
Comment (rshriram) [inline pilot/proxy/envoy/mixer.go]: why disable report at ingress? Report should have no issue right?

Comment (ayj) [reply pilot/proxy/envoy/mixer.go]: Human error. I mixed up the boolean flag usage for building opaque route config. This has been reverted and instead we disable forwarding mixer attributes to the app for sidecar inbound routes.

**Comment by ldemailly** [comment]
are the test detecting that case?

**Comment by ayj** [comment]
> are the test detecting that case?

It seems so since I forgot to refresh test data with the last round of changes and the presubmit failed.

**Comment by rshriram** [comment]
can you please fix the lint errors? Otherwise, I am disabling the bot.

File: pilot/proxy/envoy/mixer.go [hunk 4]
```diff
@@ -201,42 +207,48 @@ func mixerHTTPRouteConfig(role proxy.Node, instances []*model.ServiceInstance, c
...
 		}
 
-		apiSpecs := config.HTTPAPISpecByDestination(instance)
-		model.SortHTTPAPISpec(apiSpecs)
-		for _, config := range apiSpecs {
-			sc.HttpApiSpec = append(sc.HttpApiSpec, config.Spec.(*mccpb.HTTPAPISpec))
-		}
+		// omit API, Quota, and Auth portion of service config when
+		// check and report are disabled.
+		if !sc.DisableCheckCalls || !sc.DisableReportCalls {
```
Comment (mandarjog) [inline pilot/proxy/envoy/mixer.go]: We should not omit API/Quota/Auth when `Report` is disabled.
They are orthogonal.

Comment (ayj) [reply pilot/proxy/envoy/mixer.go]: We only omit when neither `Check` or `Report` is enabled

Comment (mandarjog) [reply pilot/proxy/envoy/mixer.go]: `sc.DisableReport` is orthogonal though. It should be only based on if `Check` is enabled.

For example, if Check if disabled, but report is enabled, we *do* want to omit.

**Comment by rshriram** [comment]
I just realized.. We don't enable the mixer filter for outbound calls. So, what is the need for this PR, unless I am missing something.

**Comment by mandarjog** [comment]
The new mixer client filter configuration (for api mgmt) did enable mixer for inbound and outbound communication. This PR fixes that.

Generally we want to support selective enablement for inbound and outbound communications shortly.

For example enabling mixer check for egress rules.

**Comment by qiwzhang** [comment]
@rshriram 
To be specifically, in the old way, mixer_filter is reading "mixer_check" and "mixer_report" from per-route opaque data for enabling check/report.   In the new way,  there is a service_configs map of destination.service to service_config.  Each service_config has its "enable_check_calls" and "enable_report_calls".  "destinaion.service" is read from per-route opaque data.   
In this new config,  "enable_check_calls" are true even for outbound traffic by mistake.  This PR tries to fix it.

**Comment by rshriram** [comment]
On Mon, Jan 8, 2018 at 1:27 PM Wayne Zhang <notifications@github.com> wrote:

> @rshriram <https://github.com/rshriram>
> To be specifically, in the old way, mixer_filter is reading "mixer_check"
> and "mixer_report" from per-route opaque data for enabling check/report. In
> the new way, there is a service_configs map of destination.service to
> service_config. Each service_config has its "enable_check_calls" and
> "enable_report_calls". "destinaion.service" is read from per-route opaque
> data.
> In this new config, "enable_check_calls" are true even for outbound
> traffic by mistake. This PR tries to fix it.
>

Nice. Thanks for the explanation!

**Comment by ldemailly** [comment]
the circle error is a compile error
```
pilot/platform/cloudfoundry/controller_test.go:31:7: undefined: gomega.NewGomegaWithT
pilot/platform/cloudfoundry/controller_test.go:31:7: too many errors
FAIL	istio.io/istio/pilot/platform/cloudfoundry [build failed]
```
unfortunately I don't think circle auto rebases for you - so can you manually rebase ?

this PR is critical for perf and we need to get it in before 0.5.0

File: pilot/proxy/envoy/testdata/lds-v1-multi-match-fault.json.golden [hunk 5]
```diff
@@ -637,16 +631,6 @@
           "quota_name": "RequestCount",
           "v2": {
            "defaultDestinationService": "hello.default.svc.cluster.local",
-           "forwardAttributes": {
-            "attributes": {
-             "source.ip": {
-              "bytesValue": "AAAAAAAAAAAAAP//CgEBAQ=="
```
Comment (ldemailly) [inline pilot/proxy/envoy/testdata/lds-v1-multi-match-fault.json.golden]: I saw those in the 0.4.0 headers, it's gone now ?

Comment (ayj) [reply pilot/proxy/envoy/testdata/lds-v1-multi-match-fault.json.golden]: Yes, I removed forwarded attributes from the list of attributes sent to the app. See https://github.com/istio/istio/pull/2419/files#diff-83699eda922e729afb2122093ca240b4R171

**Pull request merged**